File: problems/problemset/virtual-filesystem/Python/solution/vfs.py
class VFS:
    class _Node:
        __slots__ = ('is_dir', 'content', 'children')
        def __init__(self, is_dir):
            self.is_dir = is_dir
            self.content = ''
            self.children = {}

    def __init__(self):
        self._root = VFS._Node(True)

    @staticmethod
    def _split(path):
        return [p for p in path.split('/') if p]

    def _resolve(self, path):
        cur = self._root
        for p in self._split(path):
            if not cur.is_dir: return None
            cur = cur.children.get(p)
            if cur is None: return None
        return cur

    def _resolve_parent(self, path):
        parts = self._split(path)
        if not parts: return None, ''
        cur = self._root
        for p in parts[:-1]:
            if not cur.is_dir: return None, ''
            cur = cur.children.get(p)
            if cur is None: return None, ''
        return cur, parts[-1]

    @classmethod
    def _deep_copy(cls, node):
        n = cls._Node(node.is_dir)
        n.content = node.content
        for k, v in node.children.items():
            n.children[k] = cls._deep_copy(v)
        return n

    def _collect_find(self, node, base, name, out):
        for child_name, child in node.children.items():
            child_path = base + '/' + child_name
            if child_name == name:
                out.append(child_path)
            if child.is_dir:
                self._collect_find(child, child_path, name, out)

    def mkdir(self, path):
        parent, name = self._resolve_parent(path)
        if parent is None: return 'ERROR: not found\n'
        if not parent.is_dir: return 'ERROR: not a directory\n'
        if name in parent.children: return 'ERROR: already exists\n'
        parent.children[name] = VFS._Node(True)
        return ''

    def touch(self, path):
        parent, name = self._resolve_parent(path)
        if parent is None: return 'ERROR: not found\n'
        if not parent.is_dir: return 'ERROR: not a directory\n'
        if name in parent.children: return 'ERROR: already exists\n'
        parent.children[name] = VFS._Node(False)
        return ''

    def write(self, path, content):
        parent, name = self._resolve_parent(path)
        if parent is None: return 'ERROR: not found\n'
        if not parent.is_dir: return 'ERROR: not a directory\n'
        if name in parent.children:
            if parent.children[name].is_dir: return 'ERROR: is a directory\n'
            parent.children[name].content = content
        else:
            node = VFS._Node(False)
            node.content = content
            parent.children[name] = node
        return ''

    def read(self, path):
        node = self._resolve(path)
        if node is None: return 'ERROR: not found\n'
        if node.is_dir: return 'ERROR: is a directory\n'
        return node.content + '\n'

    def ls(self, path):
        node = self._resolve(path)
        if node is None: return 'ERROR: not found\n'
        if not node.is_dir: return 'ERROR: not a directory\n'
        return ''.join(k + ('/' if v.is_dir else '') + '\n'
                       for k, v in sorted(node.children.items()))

    def rm(self, path):
        parent, name = self._resolve_parent(path)
        if parent is None: return 'ERROR: not found\n'
        if name not in parent.children: return 'ERROR: not found\n'
        node = parent.children[name]
        if node.is_dir and node.children: return 'ERROR: not empty\n'
        del parent.children[name]
        return ''

    def rmr(self, path):
        parent, name = self._resolve_parent(path)
        if parent is None: return 'ERROR: not found\n'
        if name not in parent.children: return 'ERROR: not found\n'
        del parent.children[name]
        return ''

    def mv(self, src, dst):
        src_parent, src_name = self._resolve_parent(src)
        if src_parent is None: return 'ERROR: not found\n'
        if src_name not in src_parent.children: return 'ERROR: not found\n'
        dst_parent, dst_name = self._resolve_parent(dst)
        if dst_parent is None: return 'ERROR: not found\n'
        if not dst_parent.is_dir: return 'ERROR: not a directory\n'
        if dst_name in dst_parent.children: return 'ERROR: already exists\n'
        dst_parent.children[dst_name] = src_parent.children.pop(src_name)
        return ''

    def cp(self, src, dst):
        src_node = self._resolve(src)
        if src_node is None: return 'ERROR: not found\n'
        dst_parent, dst_name = self._resolve_parent(dst)
        if dst_parent is None: return 'ERROR: not found\n'
        if not dst_parent.is_dir: return 'ERROR: not a directory\n'
        if dst_name in dst_parent.children: return 'ERROR: already exists\n'
        dst_parent.children[dst_name] = self._deep_copy(src_node)
        return ''

    def find(self, path, name):
        node = self._resolve(path)
        if node is None: return 'ERROR: not found\n'
        if not node.is_dir: return 'ERROR: not a directory\n'
        base = '' if path == '/' else path
        results = []
        self._collect_find(node, base, name, results)
        results.sort()
        return ''.join(r + '\n' for r in results)

File: problems/problemset/virtual-filesystem/Rust/solution/vfs.rs
use std::collections::BTreeMap;

struct Node {
    is_dir: bool,
    content: String,
    children: BTreeMap<String, Box<Node>>,
}

impl Node {
    fn dir() -> Box<Self> {
        Box::new(Node { is_dir: true, content: String::new(), children: BTreeMap::new() })
    }
    fn file() -> Box<Self> {
        Box::new(Node { is_dir: false, content: String::new(), children: BTreeMap::new() })
    }
    fn deep_copy(&self) -> Box<Self> {
        let mut n = Box::new(Node { is_dir: self.is_dir, content: self.content.clone(), children: BTreeMap::new() });
        for (k, v) in &self.children {
            n.children.insert(k.clone(), v.deep_copy());
        }
        n
    }
}

fn split_path(path: &str) -> Vec<&str> {
    path.split('/').filter(|s| !s.is_empty()).collect()
}

fn collect_find(node: &Node, base: &str, name: &str, out: &mut Vec<String>) {
    for (child_name, child) in &node.children {
        let child_path = format!("{}/{}", base, child_name);
        if child_name == name { out.push(child_path.clone()); }
        if child.is_dir { collect_find(child, &child_path, name, out); }
    }
}

pub struct VFS {
    root: Box<Node>,
}

impl VFS {
    pub fn new() -> Self {
        VFS { root: Node::dir() }
    }

    fn resolve(&self, path: &str) -> Option<&Node> {
        let mut cur: &Node = &self.root;
        for part in split_path(path) {
            if !cur.is_dir { return None; }
            cur = cur.children.get(part)?;
        }
        Some(cur)
    }

    fn navigate_to_parent<'a>(cur: &'a mut Node, parts: &[&str]) -> Option<&'a mut Node> {
        let mut node = cur;
        for &part in parts {
            if !node.is_dir { return None; }
            node = node.children.get_mut(part)?;
        }
        Some(node)
    }

    fn extract_node(&mut self, path: &str) -> Option<Box<Node>> {
        let parts = split_path(path);
        if parts.is_empty() { return None; }
        let n = parts.len();
        let parent = Self::navigate_to_parent(&mut self.root, &parts[..n-1])?;
        parent.children.remove(parts[n-1])
    }

    fn insert_node(&mut self, path: &str, node: Box<Node>) -> String {
        let parts = split_path(path);
        if parts.is_empty() { return "ERROR: not found\n".into(); }
        let n = parts.len();
        let parent = match Self::navigate_to_parent(&mut self.root, &parts[..n-1]) {
            None => return "ERROR: not found\n".into(),
            Some(p) => p,
        };
        let name = parts[n-1];
        if !parent.is_dir { return "ERROR: not a directory\n".into(); }
        if parent.children.contains_key(name) { return "ERROR: already exists\n".into(); }
        parent.children.insert(name.to_string(), node);
        String::new()
    }

    pub fn mkdir(&mut self, path: &str) -> String {
        let parts = split_path(path);
        if parts.is_empty() { return "ERROR: not found\n".into(); }
        let n = parts.len();
        let parent = match Self::navigate_to_parent(&mut self.root, &parts[..n-1]) {
            None => return "ERROR: not found\n".into(),
            Some(p) => p,
        };
        let name = parts[n-1];
        if !parent.is_dir { return "ERROR: not a directory\n".into(); }
        if parent.children.contains_key(name) { return "ERROR: already exists\n".into(); }
        parent.children.insert(name.to_string(), Node::dir());
        String::new()
    }

    pub fn touch(&mut self, path: &str) -> String {
        let parts = split_path(path);
        if parts.is_empty() { return "ERROR: not found\n".into(); }
        let n = parts.len();
        let parent = match Self::navigate_to_parent(&mut self.root, &parts[..n-1]) {
            None => return "ERROR: not found\n".into(),
            Some(p) => p,
        };
        let name = parts[n-1];
        if !parent.is_dir { return "ERROR: not a directory\n".into(); }
        if parent.children.contains_key(name) { return "ERROR: already exists\n".into(); }
        parent.children.insert(name.to_string(), Node::file());
        String::new()
    }

    pub fn write(&mut self, path: &str, content: &str) -> String {
        let parts = split_path(path);
        if parts.is_empty() { return "ERROR: not found\n".into(); }
        let n = parts.len();
        let parent = match Self::navigate_to_parent(&mut self.root, &parts[..n-1]) {
            None => return "ERROR: not found\n".into(),
            Some(p) => p,
        };
        let name = parts[n-1];
        if !parent.is_dir { return "ERROR: not a directory\n".into(); }
        match parent.children.get_mut(name) {
            Some(node) => {
                if node.is_dir { return "ERROR: is a directory\n".into(); }
                node.content = content.to_string();
            }
            None => {
                let mut node = Node::file();
                node.content = content.to_string();
                parent.children.insert(name.to_string(), node);
            }
        }
        String::new()
    }

    pub fn read(&self, path: &str) -> String {
        match self.resolve(path) {
            None => "ERROR: not found\n".into(),
            Some(n) if n.is_dir => "ERROR: is a directory\n".into(),
            Some(n) => n.content.clone() + "\n",
        }
    }

    pub fn ls(&self, path: &str) -> String {
        match self.resolve(path) {
            None => "ERROR: not found\n".into(),
            Some(n) if !n.is_dir => "ERROR: not a directory\n".into(),
            Some(n) => {
                let mut out = String::new();
                for (name, child) in &n.children {
                    out.push_str(name);
                    if child.is_dir { out.push('/'); }
                    out.push('\n');
                }
                out
            }
        }
    }

    pub fn rm(&mut self, path: &str) -> String {
        let parts = split_path(path);
        if parts.is_empty() { return "ERROR: not found\n".into(); }
        let n = parts.len();
        let parent = match Self::navigate_to_parent(&mut self.root, &parts[..n-1]) {
            None => return "ERROR: not found\n".into(),
            Some(p) => p,
        };
        let name = parts[n-1];
        match parent.children.get(name) {
            None => return "ERROR: not found\n".into(),
            Some(node) if node.is_dir && !node.children.is_empty() => return "ERROR: not empty\n".into(),
            _ => {}
        }
        parent.children.remove(name);
        String::new()
    }

    pub fn rmr(&mut self, path: &str) -> String {
        let parts = split_path(path);
        if parts.is_empty() { return "ERROR: not found\n".into(); }
        let n = parts.len();
        let parent = match Self::navigate_to_parent(&mut self.root, &parts[..n-1]) {
            None => return "ERROR: not found\n".into(),
            Some(p) => p,
        };
        let name = parts[n-1];
        if !parent.children.contains_key(name) { return "ERROR: not found\n".into(); }
        parent.children.remove(name);
        String::new()
    }

    pub fn mv(&mut self, src: &str, dst: &str) -> String {
        // Validate dst before touching src
        let dst_parts = split_path(dst);
        if dst_parts.is_empty() { return "ERROR: not found\n".into(); }
        let dst_name = dst_parts[dst_parts.len()-1].to_string();
        let dst_parent_path = format!("/{}", dst_parts[..dst_parts.len()-1].join("/"));
        {
            match self.resolve(&dst_parent_path) {
                None => return "ERROR: not found\n".into(),
                Some(p) if !p.is_dir => return "ERROR: not a directory\n".into(),
                Some(p) if p.children.contains_key(&dst_name) => return "ERROR: already exists\n".into(),
                _ => {}
            }
        }
        match self.extract_node(src) {
            None => "ERROR: not found\n".into(),
            Some(node) => self.insert_node(dst, node),
        }
    }

    pub fn cp(&mut self, src: &str, dst: &str) -> String {
        let copied = match self.resolve(src) {
            None => return "ERROR: not found\n".into(),
            Some(n) => n.deep_copy(),
        };
        self.insert_node(dst, copied)
    }

    pub fn find(&self, path: &str, name: &str) -> String {
        match self.resolve(path) {
            None => "ERROR: not found\n".into(),
            Some(n) if !n.is_dir => "ERROR: not a directory\n".into(),
            Some(n) => {
                let base = if path == "/" { "" } else { path };
                let mut results = Vec::new();
                collect_find(n, base, name, &mut results);
                results.sort();
                results.iter().map(|r| format!("{}\n", r)).collect()
            }
        }
    }
}
